[IndVarSimplify] Refactor `handleFloatingPointIV`, modernize pass (NFC)

antoniofrighetto opened the pull request:
`handleFloatingPointIV` is now abstracted out into different routines, particularly:
- `maybeFloatingPointRecurrence` which establishes whether we handle a floating-point iv recurrence;
- `tryConvertToIntegerIV` which attempts to convert the fp start, step and exit values into integer ones;
- `canonicalizeToIntegerIV` which rewrites the recurrence.

Minor opportunity to modernize the code where possible.

dtcxzyw commented on llvm/lib/Transforms/Scalar/IndVarSimplify.cpp:
```diff
@@ -198,205 +198,275 @@ static bool ConvertToSInt(const APFloat &APF, int64_t &IntVal) {
   IntegerType *Int32Ty = Type::getInt32Ty(PN->getContext());
+  auto *Incr = cast<BinaryOperator>(PN->getIncomingValue(BackEdge));
+  auto *BI = cast<BranchInst>(FPIV.Compare->user_back());
+  assert(Incr && BI);
```
`cast` never returns null result.

dtcxzyw reviewed (approved):
LGTM.

antoniofrighetto committed:
[IndVarSimplify] Refactor `handleFloatingPointIV`, modernize pass (NFC)

`handleFloatingPointIV` is now abstracted out into different routines,
particularly:
- `maybeFloatingPointRecurrence` which establishes whether we handle a
  floating-point iv recurrence;
- `tryConvertToIntegerIV` which attempts to convert the fp start, step
  and exit values into integer ones;
- `canonicalizeToIntegerIV` which rewrites the recurrence.

Minor opportunity to modernize the code where possible.

fhahn reviewed (commented):
It looks like this is not NFC, seeing a crash for the IR below

```
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "arm64-apple-macosx15.0.0"

define ptr @old_main() {
entry:
  br label %while.body

while.body:                                       ; preds = %while.body, %entry
  %tnow.0279 = phi double [ 0.000000e+00, %entry ], [ %add168, %while.body ]
  %add168 = fadd double %tnow.0279, 0.000000e+00
  %cmp164 = fcmp olt double %add168, 0.000000e+00
  %0 = select i1 %cmp164, i1 false, i1 false
  br label %while.body
}
```
